src/v/raft/buffered_protocol.cc

```diff
@@ -0,0 +1,397 @@
+                it,
+                target_node,
+                std::make_unique<internal::append_entries_queue>(
+                  target_node,
+                  _base_protocol,
+                  _gate.hold(),
+                  _max_inflight_requests,
+                  _max_buffered_bytes));
+          }
+          return it->second->append_entries(std::move(req), std::move(opts));
+      });
+};
+
+ss::future<result<heartbeat_reply>> buffered_protocol::heartbeat(
+  model::node_id target_node, heartbeat_request&& req, rpc::client_opts opts) {
+    return apply_with_gate(
+      _gate,
+      _base_protocol,
+      target_node,
+      std::move(req),
+      std::move(opts),
+      &consensus_client_protocol::heartbeat);
+}
+
+ss::future<result<heartbeat_reply_v2>> buffered_protocol::heartbeat_v2(
+  model::node_id target_node,
+  heartbeat_request_v2&& req,
+  rpc::client_opts opts) {
+    return apply_with_gate(
+      _gate,
+      _base_protocol,
+      target_node,
+      std::move(req),
+      std::move(opts),
+      &consensus_client_protocol::heartbeat_v2);
+}
+
+ss::future<result<install_snapshot_reply>> buffered_protocol::install_snapshot(
+  model::node_id target_node,
+  install_snapshot_request&& req,
+  rpc::client_opts opts) {
+    return apply_with_gate(
+      _gate,
+      _base_protocol,
+      target_node,
+      std::move(req),
+      std::move(opts),
+      &consensus_client_protocol::install_snapshot);
+}
+
+ss::future<result<timeout_now_reply>> buffered_protocol::timeout_now(
+  model::node_id target_node,
+  timeout_now_request&& req,
+  rpc::client_opts opts) {
+    return apply_with_gate(
+      _gate,
+      _base_protocol,
+      target_node,
+      std::move(req),
+      std::move(opts),
+      &consensus_client_protocol::timeout_now);
+}
+
+ss::future<result<transfer_leadership_reply>>
+buffered_protocol::transfer_leadership(
+  model::node_id target_node,
+  transfer_leadership_request&& req,
+  rpc::client_opts opts) {
+    return apply_with_gate(
+      _gate,
+      _base_protocol,
+      target_node,
+      std::move(req),
+      std::move(opts),
+      &consensus_client_protocol::transfer_leadership);
+}
+
+ss::future<bool> buffered_protocol::ensure_disconnect(model::node_id node_id) {
+    return _base_protocol.ensure_disconnect(node_id);
+}
+
+ss::future<> buffered_protocol::reset_backoff(model::node_id node_id) {
+    return _base_protocol.reset_backoff(node_id);
+}
+
+ss::future<> buffered_protocol::stop() {
+    vlog(raftlog.debug, "stopping buffered protocol");
+    _gc_timer.cancel();
+    auto f = _gate.close();
+    co_await ss::parallel_for_each(
+      _append_entries_queues,
+      [](decltype(_append_entries_queues)::value_type& queue) {
+          return queue.second->stop();
+      });
+    _append_entries_queues.clear();
+    co_await std::move(f);
+    vlog(raftlog.debug, "stopped buffered protocol");
+}
+
+void buffered_protocol::garbage_collect_unused_queues() {
```

Comment:
why do we need this whole machinery to gc unused queues, we have one per broker per shard, doesn't seem like a lot of overhead.

Reply:
i didn't wanted to add a notification from the membership layer when the node is removed from the cluster.

Reply:
I added the garbage collection to make the queue lifecycle management less complicated that the connection management. Currently the RPC connections that Raft uses are managed outside of `raft` submodule. They are bound to the cluster membership (i.e. when node is no longer a cluster member and has been removed from raft0 the connection can be removed.) I wanted to keep the queues management simple and added this GC logic. 
Creating and removing a queue is not expensive operation.

The gc operation uses `std::swap` and is not async, therefore it is safe to remove the queue in the background.

Reply:
Thanks, makes sense.